docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+
+```
+#define SGX_IOC_ENCLAVE_MPROTECT _IOW(SGX_MAGIC, 0x06, struct sgx_enclave_mprotect)
+/**
+ * struct sgx_enclave_mprotect - parameter structure for the
+ *                                %SGX_IOC_ENCLAVE_MPROTECT ioctl
+ * @addr:	address of the memory to change protections
+ * @length:	length of the area.  This must be a multiple of the page size.
+ * @prot:	this must be or'ed of following:
+             PROT_READ
+             PROT_WRITE
+             PROT_EXEC
+             PROT_TRIM (new): change the page type to PT_TRIM, implies RW. User space should immediately EACCEPT, and then optionally call mprotect with PROT_EACCEPT.
+             PROT_TCS (new): change the page type to PT_TCS
+             PROT_EACCEPT (new): notify kernel EACCEPT is done for PT_TRIM type changes. This can be optional. Kernel will try EREMOVE on PT_TRIM pages in best effort on its own if user space does not notify EACCEPT.
+ */
+struct sgx_enclave_mprotect {
+	__u64	addr;
+	__u64	length;
+	__u64	prot;
+};
+```
+
+### Remarks
+
+Kernel should ensure that SGX instructions can succeed or catch and handle any fault.
+  - The kernel may maintain EPCM information on each page which includes access permission RWX, page types of PT_REG, PT_TRIM, PT_TCS.
+  - The kernel catches fault on EMODPR, EMODT and converts to error code returned to user space.
+
+The enclave run-time (or trusted run-time) may implement a parallel memory management structure which would provide information to the enclave on the enclave memory mappings.  The run-time can have a trusted API analogous to mmap which makes a call out of the enclave to issue the mmap  and then either perform EACCEPT on the pages and update the internal memory structures or configure the enclave to perform the EACCEPT when a #PF is delivered to the enclave.  With the outlined kernel interface, either implementation is possible.
+
+## Example advanced flows
+
+More advanced flows can be implemented as combinations of the basic flows. Here we present a few examples. 
+
+### Dynamic code loading
+
+An enclave can load trusted code to a new EPC page with execution permission using EACCEPT_COPY with following steps:
+1. Enclave calls mmap for a region in enclave ELRANGE for EAUG
+2. Enclave issues EACCEPT_COPY to copy trusted code from an existing EPC page to an unavailable page at an address inside that region
+   - Mechanism for enclave to establish the trustworthiness of new code is out-of-scope here.
+   - This results a #PF
+3. Kernel EAUG the page on #PF, set RW permissions in both EPCM and PTE
+4. The fault handler returns,  EACCEPT_COPY instruction retried, which sets RX permissions in EPCM specified in PageInfo operand.
+5. Enclave makes ocall which invokes mprotect syscall to change PTE permissions from RW to RX 
+
+### Lazy dynamic stack expansion
+An enclave can lazily expand its stacks as follows.
+1. Enclave calls mmap with MAP_GROWSDOWN for a stack region in enclave ELRANGE
```

Comment:
IIUC, with MAP_GROWSDOWN set, the kernel expands the VMA downward upon a #PF within 64KB below its lower bound. In other words, MAP_GROWSDOWN takes effect only for #PFs outside of any VMA (i.e., #PF address must be unmapped). It doesn't sound like we are copying that behavior to SGX. What exactly does `MAP_GROWSDOWN` do in the context of SGX? Or what's the difference in behaviors between a VMA with `MAP_GROWSDOWN` set and a VMA without it?

Reply:
@binxing and @dimakuv, those are good points. 
The desired behavior is for kernel to EAUG all pages between the current low end of the populated region and the faulting address if an VMA is marked with GROWSDOWN, and faulting address is below the low end of populated.
However, I also found that MAP_GROWSDOWN is not allowed for MAP_SHARED and/or fd backed mapping. So this may require some discussion in kernel community to change the current behavior, or we need come up with a different flag.

Reply:
>However, I also found that MAP_GROWSDOWN is not allowed for MAP_SHARED and/or fd backed mapping. 

I didn't look at the current Linux code but it looks like:
1. Combination of `MAP_GROWSDOWN | MAP_SHARED` is legal (if there is no fd backing, i.e. if there is also `MAP_ANONYMOUS`).
2. Combination of `MAP_GROWSDOWN` and a backing fd is illegal. Intuitively it makes sense -- `MAP_GROWSDOWN` is purely for stacks, and a file-backed stack sounds like nonsense.

But going back to the original flow: does the Linux kernel even need to know about `MAP_GROWSDOWN`? This can be fully emulated by the enclave runtime + Intel SGX driver, I believe. Probably there will be a need for a new mprotect flag/ioctl to pass the enclave's intention of growing the enclave stack to the SGX driver... But we already do similar things to circumvent `munmap()` (I mean `PROT_TRIM`).

Also, I think this "lazy dynamic stack extension" is a very rare use case for SGX enclaves. We haven't observed any Graphene applications in the wild that wanted to extend the stack at runtime (we simply allocate a large-enough stack at runtime, and it always works).

Reply:
Agreed.